B2B IPC

pawalt opened the pull request:
This PR adds the ability for our backends to securely communicate with one another by:
- Adding attest, refresh, and JWKs views for managing JWTs
  - Where the attest view uses DOT credentials for authentication
- Adding tests for those views

A separate PR will be made for DLA to provide easy wrappers for b2b ipc.
[ch1738]

KeenWill commented on Platform/settings/base.py:
```diff
@@ -25,6 +25,21 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = os.environ.get("SECRET_KEY", "o7ql0!vuk0%rgrh9p2bihq#pege$qqlm@zo#8&t==%&za33m*2")
+IDENTITY_RSA_PRIVATE_KEY = os.environ.get("IDENTITY_RSA_PRIVATE_KEY", """-----BEGIN RSA PRIVATE KEY-----
```
hello secret key?

davish replied:
lmao

pawalt replied:
it's a dummy key big guy look at the environment substitution

KeenWill replied:
lel im blind

KeenWill commented:
@pawalt writing tests? Is the world ending?

ArmaanT reviewed (changes requested):
These are some surface-level comments. I'll dive deeper in a bit

pawalt commented on identity/views.py:
```diff
@@ -0,0 +1,146 @@
+            return JsonResponse(data={"error": "unauthenticated"}, status=HTTPStatus.UNAUTHORIZED)
+
+        # pulls out name as recorded in DOT application database
+        local_name = slugify(request.client.name)
```
slugify will do the lower

pawalt commented on identity/views.py:
```diff
@@ -0,0 +1,146 @@
+        request.client = None
+        authenticated = self.validator.authenticate_client(request, *args, **kwargs)
+        if not authenticated:
+            return JsonResponse(data={"error": "unauthenticated"}, status=HTTPStatus.UNAUTHORIZED)
```
It's good practice to use the enums just to reduce the magic-number-ness

pawalt commented on identity/views.py:
```diff
@@ -0,0 +1,146 @@
+    """
+
+    # building out JWKS view at init time so we don't have to recalculate it for each request
+    def __init__(self, **kwargs):
```
yeah its almost exactly the same

pawalt commented on identity/views.py:
```diff
@@ -0,0 +1,146 @@
+            refresh_jwt = jwt.JWT(key=id_privkey, jwt=refresh_jwt_raw)
+            claims = json.loads(refresh_jwt.claims)
+            if "use" not in claims or claims["use"] != "refresh":
+                raise Exception("expected JWT with `use -> refresh` claim")
```
I'm raising the exception so it can be caught to serve a jsonresponse

ArmaanT replied:
Is there a reason for the exception then jsonresponse in an except block? If The ony thing that will fail is not using a refresh token, I think we should just inline the JsonResponse for it

pawalt commented on identity/views.py:
```diff
@@ -0,0 +1,146 @@
+- exp -> expiry_time - this makes sure our JWT is only valid for EXPIRY_TIME
+"""
+def mint_access_jwt(key: jwk.JWK, urn: str) -> jwt.JWT:
+    now = time.time()
```
Why does it need to be? I want the time to be unix time so that it doesn't need timezones. This is part of the JWT spec fwiw

pawalt commented on Platform/settings/base.py:
```diff
@@ -25,6 +25,21 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = os.environ.get("SECRET_KEY", "o7ql0!vuk0%rgrh9p2bihq#pege$qqlm@zo#8&t==%&za33m*2")
+IDENTITY_RSA_PRIVATE_KEY = os.environ.get("IDENTITY_RSA_PRIVATE_KEY", """-----BEGIN RSA PRIVATE KEY-----
```
hows this look?

pawalt commented on Platform/settings/base.py:
```diff
@@ -25,6 +25,24 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = os.environ.get("SECRET_KEY", "o7ql0!vuk0%rgrh9p2bihq#pege$qqlm@zo#8&t==%&za33m*2")
+IDENTITY_RSA_PRIVATE_KEY = os.environ.get(
```
dup

ArmaanT commented on identity/utils.py:
```diff
@@ -0,0 +1,39 @@
+
+def mint_access_jwt(key: jwk.JWK, urn: str) -> jwt.JWT:
+    """
+    Minting JWTs with the following claims:
```
change to `Mint a JWT with the following claims:` and make similar changes to the rest of the docstring. (same with refresh)
the imperative mood is a bit easier to read

pawalt replied:
good call

ArmaanT commented on identity/views.py:
```diff
@@ -0,0 +1,146 @@
+    Implements an endpoint to attest with client id + client secret
+
+    This endpoint returns a refresh JWT with an unlimited lifetime and an access JWT with
+    a short lifetime.
```
Still missing the link here

ArmaanT commented on identity/views.py:
```diff
@@ -0,0 +1,102 @@
+        return self.jwks_response
+
+
+class RefreshJWTView(View):
```
Do you think it makes sense to use DRF's ApiView as a base for these instead of django's View?

pawalt replied:
I don't think it makes sense to draw a dependency on DRF just for that.

ArmaanT replied:
I looked into this more and `ApiView` doesn't provide much additional value, no need to use it then.

ArmaanT reviewed (changes requested):
Few more comments

pawalt commented on identity/utils.py:
```diff
@@ -0,0 +1,39 @@
+EXPIRY_TIME = 15 * 60  # 15 minutes
+
+
+def mint_access_jwt(key: jwk.JWK, urn: str) -> jwt.JWT:
```
thanks dad :)

pawalt commented:
@ArmaanT You commented in some weird way that I can't respond to so I'll respond here to leftover comments:

> Still missing the link here

I think you're looking at the wrong place. I used the DOT code for the JWKS view. I've linked there.

> Is there a reason for the exception then jsonresponse in an except block? If The ony thing that will fail is not using a refresh token, I think we should just inline the JsonResponse for it

I found out there's a JWException for jwt validation errors, so I've moved to and switched to a JSONResponse there as you suggested.

ArmaanT commented on identity/utils.py:
```diff
@@ -0,0 +1,41 @@
+
+SIGNING_ALG = "RS256"
+EXPIRY_TIME = 15 * 60  # 15 minutes
+id_privkey = jwk.JWK.from_pem(settings.IDENTITY_RSA_PRIVATE_KEY.encode("utf-8"))
```
Super nit, but I think this should be in all caps since it's technically a constant, right?

pawalt replied:
yeah

ArmaanT reviewed (changes requested):
A couple of very minor comments. The three things that I want to check out a bit more are:
* using django's timezone aware utils to get the current time
* Figuring out why the code directly pulled from DOT for oauth2 validation isn't working here
* We definiely should set this PR up in a staging environment (even if it's done manually) and test with the DLA changes (that will need to be written) to iron out the kinks.
Other than that, this is looking really good!

pawalt commented on identity/views.py:
```diff
@@ -0,0 +1,102 @@
+    View used for refreshing access JWTs
+    """
+
+    def get(self, request, *args, **kwargs):
```
I was thinking GET because it has no body and is idempotent

vsingh18567 reviewed (commented):
Just one question regarding the `AttestView`

ArmaanT commented on identity/views.py:
```diff
@@ -0,0 +1,110 @@
+            return JsonResponse(data={"error": "unauthenticated"}, status=HTTPStatus.UNAUTHORIZED)
+
+        # pulls out name as recorded in DOT application database
+        local_name = slugify(request.client.name)
```
That gets set on line 35. One of the things the validator does is set `request.client` to be the instance of the OAuth2 Application making the request.

ArmaanT commented on tests/identity/test_views.py:
```diff
@@ -0,0 +1,174 @@
+        self.assertEqual(expected_urn, refresh_claims["sub"])
+        self.assertEqual("access", access_claims["use"])
+        self.assertEqual("refresh", refresh_claims["use"])
+        now = time.time()
```
Yep, it would be slightly different, but we don't do an exact check on the time, just greater than/less than checks

ArmaanT commented on tests/identity/test_views.py:
```diff
@@ -0,0 +1,174 @@
+        response = self.client.post(reverse("identity:attest"), **auth_headers)
+        content = response.json()
+        self.assertIsInstance(content, dict)
+        self.assertEqual(response.status_code, HTTPStatus.UNAUTHORIZED)
```
Discussed in person. We probably don't need to verify the existence of an error message since the 401 means we hit an error response.

ArmaanT commented on tests/identity/test_views.py:
```diff
@@ -0,0 +1,174 @@
+        response = self.client.post(reverse("identity:attest"), **auth_headers)
+        content = response.json()
+        self.assertIsInstance(content, dict)
+        self.assertEqual(response.status_code, HTTPStatus.UNAUTHORIZED)
```
also same as above
